File: leetcode/383.ransom-note.py
# @leet start
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        CODE_POINT_A = ord("a")
        counts = [0] * 26

        for ch in magazine:
            counts[ord(ch) - CODE_POINT_A] += 1

        for ch in ransomNote:
            if not counts[ord(ch) - CODE_POINT_A]:
                return False

            counts[ord(ch) - CODE_POINT_A] -= 1

        return True


# @leet end

File: leetcode/383.ransom-note.ts
// @leet start
function canConstruct(ransomNote: string, magazine: string): boolean {
  if (ransomNote.length > magazine.length) {
    return false;
  }

  const counts: Map<string, number> = new Map();

  for (const c of magazine) {
    counts.set(c, (counts.get(c) ?? 0) + 1);
  }

  for (const c of ransomNote) {
    if (!counts.has(c)) {
      return false;
    }

    const newCount = counts.get(c)! - 1;
    counts.set(c, newCount);

    if (!newCount) {
      counts.delete(c);
    }
  }

  return true;
}
// @leet end
